batcher/sweep_batch.go

```diff
@@ -0,0 +1,1227 @@
+}
+
+// Run is the batch's main event loop.
+func (b *Batch) Run(ctx context.Context) error {
+	// Dispatch the blockheight monitor.
+	go b.monitorBlockHeight(ctx)
+
+	// Dispatch the spend monitor.
+	go b.monitorSpend(ctx)
+
+	// We use a timer chan in order to trigger the batch publishment a few
+	// seconds after a new block is received. This is done avoid publishing
+	// the transaction on the same block in which it was spent. By waiting,
+	// we give the tiny amount of time needed by the spend monitor to
+	// receive the notification and close the batch, preventing further
+	// publishments.
+	var timerChan <-chan time.Time
+
+	b.log.Infof("started, primary %x, total sweeps %v",
+		b.PrimarySweepID[0:6], len(b.Sweeps))
+
+	for {
+		select {
+		case sweep := <-b.RequestChan:
+			err := b.addSweep(ctx, *sweep)
+			if err != nil {
+				return err
+			}
+
+		case height := <-b.blockEpochChan:
+			b.log.Debugf("received block %v", height)
+
+			firstHeight := (b.currentHeight == 0)
+			b.currentHeight = height
+
+			// We skip the step if this was the first height we
+			// received, as this means we just started running. This
+			// prevents immediately publishing a transaction on
+			// batch creation.
+			if !firstHeight {
+				timerChan = time.After(batchPublishDelay)
+			} else {
+				b.log.Debugf("skipping step on first height")
+			}
+		case <-timerChan:
+			err := b.step(ctx)
+			if err != nil {
+				return err
+			}
+
+		case spend := <-b.spendChan:
+			err := b.handleSpend(ctx, spend.SpendingTx)
+			if err != nil {
+				return err
+			}
+
+		case <-b.confChan:
+			return b.handleConf(ctx)
+
+		case err := <-b.errChan:
+			return err
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// step executes the next step in the batch's lifecycle.
+func (b *Batch) step(ctx context.Context) error {
+	switch b.State {
+	// For an open state we want to publish the latest version of the batch
+	// transaction. This should contain the latest set of sweeps and the
+	// most up-to-date fee rate at current blockheight.
+	case Open:
+		err := b.publish(ctx)
+		if err != nil {
+			return err
+		}
+
+	// For a closed state we want to dispatch a fresh confirmation monitor.
+	// The reason this is done in the step function, therefore on each new
+	// block, is to ensure that the batch is always monitoring the latest
+	// version of the batch transaction. This could prove useful on restarts
+	// or on re-org events.
+	case Closed:
+		go b.monitorConfirmations(ctx)
+	}
+
+	return nil
+}
+
+// publish creates and publishes the latest batch transaction to the network.
+func (b *Batch) publish(ctx context.Context) error {
+	go b.monitorSpend(ctx)
+
+	var (
+		err         error
+		fee         btcutil.Amount
+		coopSuccess bool
+	)
+
+	// Run the RBF rate update.
+	err = b.updateRbfRate(ctx)
+	if err != nil {
+		return err
+	}
+
+	fee, err, coopSuccess = b.publishBatchCoop(ctx)
+	if err != nil {
+		b.log.Warnf("co-op publish error: %v", err)
```

Comment:
IIUC this is problematic as if we publish, then republish and fail to replace then we'll immediately fall back to script spend. Parsing the error can help avoid this issue.

Reply:
Hm, nice observation.
I wouldn't say this is ideal, although the coop-tx is smaller in size compared to uncoop-tx, so since the feerate is the same then if coop failed to rbf, so will uncoop

but yeah I will try to find a way to signal that it was an rbf failure (maybe search for text in err message)

Reply:
This only falls back to non-coop if server is unresponsive.
The `coopSuccess` serves that purpose. If we get to the point where the server coop signs with MuSig2 we never fall back to non-coop spending

Reply:
> This only falls back to non-coop if server is unresponsive. The `coopSuccess` serves that purpose. If we get to the point where the server coop signs with MuSig2 we never fall back to non-coop spending

Indeed, missed that. OK, sgtm.